pkg/controller/worker/state_reconciler.go

```diff
@@ -0,0 +1,126 @@
+	"github.com/gardener/gardener-extensions/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+
+	"github.com/go-logr/logr"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+)
+
+const (
+	// StartToSyncState is used as part of the Event 'reason' when a Worker state starts to synchronize
+	StartToSyncState = "SynchronizingState"
+	// SuccessSynced is used as part of the Event 'reason' when a Worker state is synced
+	SuccessSynced = "StateSynced"
+	// ErrorStateSync is used as part of the Event 'reason' when a Worker state fail to sync
+	ErrorStateSync = "ErrorSynchronizingState"
+	// StateSyncControllerName is the name of the controller which synchronize the Worker state
+	StateSyncControllerName = "worker-state-cotroller"
+)
+
+type stateReconciler struct {
+	logger   logr.Logger
+	actuator StateActuator
+
+	// Recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	recorder record.EventRecorder
+
+	ctx    context.Context
+	client client.Client
+}
+
+// NewStateReconciler creates a new reconcile.Reconciler that reconciles
+// Worker's State resources of Gardener's `extensions.gardener.cloud` API group.
+func NewStateReconciler(mgr manager.Manager, actuator StateActuator) reconcile.Reconciler {
+	return &stateReconciler{
+		logger:   log.Log.WithName(StateUpdatingControllerName),
+		actuator: actuator,
+		recorder: mgr.GetEventRecorderFor(StateSyncControllerName),
+	}
+}
+
+func (r *stateReconciler) InjectFunc(f inject.Func) error {
+	return f(r.actuator)
+}
+
+func (r *stateReconciler) InjectClient(client client.Client) error {
+	r.client = client
+	return nil
+}
+
+func (r *stateReconciler) InjectStopChannel(stopCh <-chan struct{}) error {
+	r.ctx = util.ContextFromStopChannel(stopCh)
+	return nil
+}
+
+func (r *stateReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	worker := &extensionsv1alpha1.Worker{}
+	if err := r.client.Get(r.ctx, request.NamespacedName, worker); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Deletion flow
+	if worker.DeletionTimestamp != nil {
+		//Nothing to do
+		return reconcile.Result{}, nil
+	}
+
+	// Reconcile flow
+	operationType := gardencorev1beta1helper.ComputeOperationType(worker.ObjectMeta, worker.Status.LastOperation)
+	if operationType != gardencorev1beta1.LastOperationTypeReconcile {
+		return reconcile.Result{Requeue: true}, nil
```

Comment:
In case create there can be deployed a lot of resources which are going to trigger  reconciliation multiple times.
During this period the `Machines` have no `Status.Node` until the end when worker.status.lastOperation is set to "Success".
So if worker haven't been created yet (even if only one `Machine is pending`) , then is no big deal to loose its state.
If I do not re-queue the request after successful creation there will be no changes to trigger status update.
And by the way I am compering `operationType` which is the current operation not the lastOperation
In case of successful creation on the next row the  `operationType` will be "Reconcile" and the status updating controller will update the status at once.